rmf_site_editor/src/site/scenario.rs

```diff
@@ -102,65 +135,214 @@ pub fn update_current_scenario(
+
+        if let Some((scenario_child, _)) = instance_entities
+            .iter()
+            .find(|(_, i)| *i == instance_entity)
+        {
+            parent_pose = scenario_entities
+                .get(*scenario_child)
+                .ok()
+                .and_then(|(i, _)| {
+                    i.pose().or(recall_instance
+                        .get(*scenario_child)
+                        .ok()
+                        .and_then(|r| r.pose))
+                });
+        }
+        entity = parent_entity;
+    }
+    parent_pose
+}
+
+/// This system current searches for scenario children entities with the Instance component
+/// TODO(@xiyuoh) generalize this at some point to use T
+pub fn get_scenario_instance_entities(
+    entity: Entity,
+    children: &Query<&Children>,
+    scenario_entities: &Query<(&mut Instance, &Affiliation<Entity>)>,
+) -> Vec<(Entity, Entity)> {
+    let mut scenario_instances = Vec::new();
+    if let Ok(scenario_children) = children.get(entity) {
+        for child in scenario_children.iter() {
+            if let Some(affiliated_entity) =
+                scenario_entities.get(*child).ok().and_then(|(_, a)| a.0)
+            {
+                scenario_instances.push((*child, affiliated_entity));
+            }
+        }
+    };
+    scenario_instances
+}
 
-                let existing_added_instance: Option<&mut (Entity, Pose)> = current_scenario
-                    .added_instances
-                    .iter_mut()
-                    .find(|(e, _)| *e == entity);
-                if let Some(existing_added_instance) = existing_added_instance {
-                    existing_added_instance.1 = pose.clone();
-                    return;
-                } else if pose.is_added() {
-                    current_scenario
-                        .added_instances
-                        .push((entity, pose.clone()));
-                    return;
+/// Checks that the current scenario's included instances are categorized correctly
+pub fn handle_instance_updates(
+    mut commands: Commands,
+    mut change_current_scenario: EventWriter<ChangeCurrentScenario>,
+    mut scenario_instances: Query<(&mut Instance, &Affiliation<Entity>)>,
+    mut update_instance: EventReader<UpdateInstance>,
+    children: Query<&Children>,
+    current_scenario: Res<CurrentScenario>,
+    scenarios: Query<(Entity, &mut Scenario<Entity>)>,
+    model_instances: Query<&NameInSite, With<InstanceMarker>>,
+    recall_instance: Query<&RecallInstance>,
+) {
+    for update in update_instance.read() {
+        let parent_pose = retrieve_parent_pose(
+            update.instance,
+            update.scenario,
+            &children,
+            &recall_instance,
+            &scenarios,
+            &scenario_instances,
+        );
+        match update.update_type {
+            UpdateInstanceType::Add(new_pose) => {
+                commands
+                    .spawn(Instance::new_added(new_pose.clone()))
```

Comment:
This feels suspicious to me. We're spawning a new instance modifier entity without checking if one already exists under this scenario for this instance. If a user repeatedly hid and added an instance to the scenario over and over again, I believe we'd be adding a new modifier entity each time.

I'm pretty sure the safer thing to do here is check whether an affiliated modifier entity already exists for this scenario before spawning a new one.

Reply:
The intention is to trigger the `UpdateInstance::Add` event only when a new model instance has been spawned and placed in the site, so it shouldn't happen when existing model instances are being toggled or modified.

I've since moved this into the new system `insert_new_instance_modifiers` with the suggested checks in [a1e5719](https://github.com/open-rmf/rmf_site/pull/266/commits/a1e5719dc3dc2b533ae8ae0f036cb3f119ff4070) using `Added<T>` instead of `is_added()` as well.